lib/puppetserver/ca/action/generate.rb

```diff
@@ -103,27 +120,68 @@ def run(input)
 
           # Load, resolve, and validate puppet config settings
           settings_overrides = {}
-          # Since puppet expects the key to be called 'dns_alt_names', we need to use that here
-          # to ensure that the overriding works correctly.
-          settings_overrides[:dns_alt_names] = input['subject-alt-names'] unless input['subject-alt-names'].empty?
           puppet = Config::Puppet.new(config_path)
           puppet.load(settings_overrides)
           return 1 if CliParsing.handle_errors(@logger, puppet.errors)
 
+          # We don't want generate to respect the alt names setting, since it is usually
+          # used to generate certs for other nodes
+          alt_names = input['subject-alt-names']
+
           # Load most secure signing digest we can for csr signing.
           signer = SigningDigest.new
           return 1 if CliParsing.handle_errors(@logger, signer.errors)
 
           # Generate and save certs and associated keys
-          all_passed = generate_certs(certnames, puppet.settings, signer.digest)
+          if input['ca-client']
+            all_passed = generate_authorized_certs(certnames, alt_names, puppet.settings, signer.digest)
+          else
+            all_passed = generate_certs(certnames, alt_names, puppet.settings, signer.digest)
+          end
           return all_passed ? 0 : 1
         end
 
+        # Certs authorized to talk to the CA API need to be signed offline,
+        # in order to securely add the special auth extension.
+        def generate_authorized_certs(certnames, alt_names, settings, digest)
+          # Make sure we have all the directories where we will be writing files
+          FileSystem.ensure_dirs([settings[:ssldir],
+                                  settings[:certdir],
+                                  settings[:privatekeydir],
+                                  settings[:publickeydir]])
+
+          ca = Puppetserver::Ca::LocalCertificateAuthority.new(digest, settings)
+          ca_cert, ca_key = ca.load_ca
+          return 1 if CliParsing.handle_errors(@logger, ca.errors)
+
+          passed = certnames.map do |certname|
+            if !alt_names.empty?
+              # When validating the cert, OpenSSL will ignore the CN field if
+              # altnames are present, so we need to ensure that the certname is
+              # also listed among the alt names.
+              alt_names += ",DNS:#{certname}"
+              alt_names = Puppetserver::Ca::Utils::Config.munge_alt_names(alt_names)
+            end
+
+            # For certs signed offline, any alt names are added directly to the cert,
+            # rather than to the CSR.
+            key, csr = generate_key_csr(certname, settings, digest)
+            return false unless csr
```

Comment:
I think this will break out of the certnames loop and directly return false from `generate_authorized_certs`?

Reply:
I think since this is in a `map` call, it just ends up populating `passed` with trues and falses. But I'll try it out.